lib/commons/integrity/check/unique_csv_identifiers.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+require_relative 'base'
+require 'csv'
+
+module Commons
+  module Integrity
+    class Check
+      # Check that any values in the given columns are unique within the file
+      #
+      # == Configuration Options
+      # * column_names: the columns containing the IDs (default: ["wikidata", "ms_fb"])
+      # * column_case: "fixed" if the column_name must be exactly as specified (default: "any")
+      class UniqueCSVIdentifiers < Base
+        # @return [Array<Error>]
+        # Errors will be in the category `:unique_csv_identifiers`
+        def errors
+          return [] unless pathname_regex.match? pathname.to_s
+          problematic_values.map do |column_name, value|
+            error(unique_csv_identifiers: "Duplicated #{column_name} value: #{value}")
+          end
+        end
+
+        private
+
+        DEFAULT_COLUMN_NAMES = %w[wikidata ms_fb].freeze
+        DEFAULT_COLUMN_CASE = 'any'
+
+        def csv
+          @csv ||= CSV.read(pathname, headers: true)
+        end
+
+        def pathname_regex
+          Regexp.new(my_config.to_h['pathname_regex'] || '.*')
+        end
+
+        def headers
+          csv.headers
+        end
+
+        def column_names
+          (my_config.to_h['column_names'] || DEFAULT_COLUMN_NAMES).map do |column_name|
+            column_name.send(comparison_conversion)
+          end
+        end
+
+        def comparison_type
+          my_config.to_h['column_case'] || DEFAULT_COLUMN_CASE
+        end
+
+        def comparison_conversion
+          return :to_s if comparison_type == 'fixed'
+          :downcase
+        end
+
+        def wikidata_columns
+          headers.select do |header|
+            column_names.include? header.send(comparison_conversion)
+          end
+        end
+
+        def wikidata_columns?
+          !wikidata_columns.all?(&:nil?)
+        end
+
+        def column_values(column_name)
+          csv.map { |row| row[column_name] }
+        end
+
+        def problematic_values_by_column(column_name)
+          problematic_values_for_column = column_values(column_name).group_by(&:itself).select do |_group, values|
+            values.length > 1
+          end.keys
+          problematic_values_for_column.map do |value|
+            [column_name, value]
+          end
```

Comment:
The data structure here was a little complicated/confusing for me to easily follow. As it's a private method, called only from one place, I suspect it might be much simpler to have it simply return the duplicate values, and have the caller take care of combining that with the column name for its purposes.

Reply:
NB: wherever you need to do this, it might be simpler to do it as `[column_name].product(problematic_values_for_column)`

Reply:
I think I've addressed this now, as suggested.